File: src/data/courses/dsa/stacks/monotonic-stack/stock-span/code/python/stock_span.py
# ============================================================
# Online Stock Span
# ============================================================

# ============================================================
# Solution 1: Brute Force
# Time: O(N) per call | Space: O(N)
# ============================================================
class StockSpanner1:
    def __init__(self):
        self.prices = []

    def next(self, price: int) -> int:
        self.prices.append(price)
        span = 1
        i = len(self.prices) - 2
        while i >= 0 and self.prices[i] <= price:
            span += 1
            i -= 1
        return span

# ============================================================
# Solution 2: Monotonic Stack (Optimal)
# Time: O(1) amortized | Space: O(N)
# ============================================================
class StockSpanner2:
    def __init__(self):
        self.stack = []  # (price, index)
        self.day = 0

    def next(self, price: int) -> int:
        # Pop all prices <= current
        while self.stack and self.stack[-1][0] <= price:
            self.stack.pop()

        # Span = distance to previous greater
        span = self.day + 1 if not self.stack else self.day - self.stack[-1][1]

        self.stack.append((price, self.day))
        self.day += 1
        return span

File: src/data/courses/dsa/stacks/monotonic-stack/stock-span/code/java/StockSpan.java
// ============================================================
// Online Stock Span
// ============================================================

import java.util.*;

// ============================================================
// Solution 1: Brute Force
// Time: O(N) per call | Space: O(N)
// ============================================================
class StockSpanner1 {
    List<Integer> prices = new ArrayList<>();

    public int next(int price) {
        prices.add(price);
        int span = 1, i = prices.size() - 2;
        while (i >= 0 && prices.get(i) <= price) { span++; i--; }
        return span;
    }
}

// ============================================================
// Solution 2: Monotonic Stack (Optimal)
// Time: O(1) amortized | Space: O(N)
// ============================================================
class StockSpanner2 {
    Deque<int[]> stack = new ArrayDeque<>(); // [price, index]
    int day = 0;

    public int next(int price) {
        while (!stack.isEmpty() && stack.peek()[0] <= price)
            stack.pop();

        int span = stack.isEmpty() ? day + 1 : day - stack.peek()[1];
        stack.push(new int[]{price, day});
        day++;
        return span;
    }
}
